Feature/trlm 3d

maddyscientist opened the pull request:
This PR adds support for finding eigenvectors of the 3-d Laplace operator.
* Add new Eigensolver algorithm `TRLM3D` which is a batched 3-d eigensolver
* Fixes support for autotuning with split grid enabled
   * When we return to the global grid, the tunecaches from each sub grid are merged with all processes having a copy of this uniform cache
   * This enables support for sub-grid tuning as opposed to world-wide tuning only as previously.
* Add support for using the resident smeared gauge field for the eigensolver operator (Laplace operator only supported at present)
* Add support for host verification of Laplace 3-d and 4-d operators when running the Eigensolver
* Fixes a bug with candidate tuning (uninitialized variable)

A couple of things are left to do:
- [x] Only broadcast and merge the split grid tunecaches if we need to
- [x] Add ctest for Laplace 3-d and 4-d eigensolves

maddyscientist committed:
First attempt at porting cpviolators laplace-3d eigensolver to modern QUDA

maddyscientist committed:
Fix the 3-d blas routines to deal with single parity fields (mostly - need to add intrinsic field parity for odd single parity fields).

cpviolator committed:
Zero out vectors in rotation, fix array name

cpviolator committed:
Add appropriate convergence and locking criteria

cpviolator committed:
Cleaner min max deduction, Chebyshev, and verbosity

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into feature/trlm_3d

maddyscientist committed:
Check that we don't use e/o preconditioning with 3-d Laplace operator

maddyscientist committed:
Fix trlm-3d for multi process: 3-d inner products were not being done correctly

cpviolator committed:
Minor clarification between conditionals active_3d and converged_3d usage

maddyscientist committed:
Optimization of tlrm-3d: remove some 4-d -> 3-d -> 4-d copying

maddyscientist committed:
Add blas3d::swap function for swapping time slices between two fields.  This avoids the need for staged copying in TRLM3D::reorder3D

maddyscientist committed:
Small cleanup

maddyscientist committed:
blas3d::caxpby and blas3d::axpby should keep their scalars in the arg structure to reduce copy latency: these kernels are now much faster

maddyscientist committed:
Work around to ensure that all processes take part, even if their local eigenvalues have converged

maddyscientist committed:
Tweak stdout from performGaugeSmearQuda

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into feature/trlm_3d

maddyscientist committed:
Some fixes to ensure 3d evalues are correctly returned to the host.  Add convergence fix from Colin

maddyscientist committed:
Set kappa correclty in test code for Laplace 3d operator

maddyscientist committed:
Add optional last argument to norm_2 function in host test code to allow for disable of global sum

maddyscientist committed:
Add new setTuning(), pushTuning(), popTuning() methods to enable/disable tuning on the fly.  Remove legacy QudaInvertParam::tune parameter and remove unneeded enum QudaTune

maddyscientist committed:
In TRLM-3d: disable autotuning for block::blas and if we have process divergence.  This allows us to run with autotuning most of the time for TRLM-3d.  Some cleanup with TRLM3D::getArrayMinMax3D

maddyscientist committed:
Initialization cleanup in TRLM3D

maddyscientist committed:
Remove missed QUDA_TUNE_YES line

maddyscientist committed:
Fix bug with TRLM-3d when partitioning space

cjmorningstar10 committed:
Minor changes to lib/eig_trlm_3d.cpp -- convergence check in TRLM3D::eigensolveFromArrowMat3D,
a few indexing fixes, and addition to output

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into feature/trlm_3d

maddyscientist committed:
Add support for in-place gauge-field smearing followed by eigensolve using the smeared gauge field.  Added parameter QudaEigParam::use_smeared_gauge to facilitate this

maddyscientist committed:
Add host verification for the Laplace (3-d or 4-d) eigenvector computation.  Staggered eigensolve test now supports inline smearing and then computing the eigenspectrum using the smeared gauge field.  Some clean up of smearing command line parameters.  Fix a problem with the solver if eig_param is not passed into the invert param structure

maddyscientist committed:
Fix issues with io_test if pure double precision build

maddyscientist committed:
Add communication functionality which will be used in next commit for improved split-grid tuning: add comm_coord_global and comm_dim_global for querying global communicator coordinates and dimensions when in split grid; add CommKey::operator==

maddyscientist committed:
Help the compiler with deducting return type

maddyscientist committed:
Autotuning now uses the local communucator and not global.  Add new function joinTuneCache which joins the tunecaches from a list of global process ids

maddyscientist committed:
When reverting from split grid to global communicators, merge the tunecaches

maddyscientist committed:
Use rank 0 for local tuning rank when joining

maddyscientist committed:
Fix a couple of bugs with the split-grid tuning.  Fix a long time uninitiatlized error with candidate tuning

maddyscientist committed:
staggered tests should be built if QUDA_DIRAC_LAPLACE=ON

maddyscientist committed:
MPI errors are no longer fatal, and handling any errors is deferred to QUDA in the case of an MPI build and to QMP in the case of a QMP build

maddyscientist committed:
TRLM-3d now uses split grid: all kernels are now autotuned, and each timeslice group is now independent

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into feature/trlm_3d

maddyscientist committed:
Fix compiler warning and switch to using array object in CommKey for increased robustness

maddyscientist committed:
Remove redundant check

maddyscientist committed:
Fix some CI warnings

maddyscientist committed:
Only broadcast the tunecache when returning to global  communicator if the tunecache has changed in any sub-grid

hummingtree commented on include/quda.h:
```diff
@@ -506,6 +505,10 @@ extern "C" {
 
+    /** Whether to use the smeared gauge field for the Dirac operator
+        for whose eigenvalues are are computing. */
+    bool use_smeared_gauge;
```
Why ... don't we use `QudaBoolean`?

maddyscientist replied:
Eventually I want to deprecate and remove `QudaBoolean` from QUDA completely.  `bool` type was introduced with C99, and it's safe to assume that it is always present which wasn't necessarily the case 15 years ago when we started QUDA.

hummingtree commented on lib/communicator_stack.cpp:
```diff
@@ -49,8 +50,16 @@ namespace quda
     if (comm_nvshmem_enabled())
-      errorQuda(
-        "Split-grid is currently not supported with NVSHMEM. Please set QUDA_ENABLE_NVSHMEM=0 to disable NVSHMEM.");
+      errorQuda("Split-grid is currently not supported with NVSHMEM. Set QUDA_ENABLE_NVSHMEM=0 to disable NVSHMEM.");
```
Just ... curious, why removing the 'Please'?

maddyscientist replied:
So clang-format would fit it on one line 😉

hummingtree reviewed (approved):
The issue regarding having tuning with split grid is solved cleanly! I just have two trivial questions :)

maddyscientist committed:
Switch to using cuda/atomic header instead of cuda/std/atomic

maddyscientist committed:
Fix uninitalized valgrind warning with QudaEigParam::use_smeared_gauge

maddyscientist committed:
split grid ctests are now always enabled

maddyscientist committed:
Fix bug with fused Mobius kernel when using multiple RHS

maddyscientist commented:
I have pushed a few more updates:
* split grid tests are now always run (previously only run if tuning disabled)
* fixed a bug with MRHS fused Mobius Dslash I found while doing additional testing of this branch
* fixed an uninitialized variable (`QudaEigParam::use_smeared_gauge` added to check_param.h)

weinbe2 commented on include/kernels/blas_3d.cuh:
```diff
@@ -0,0 +1,341 @@
+    __device__ __host__ double init() const { return double(); }
+  };
+
+  template <int reduction_dim, class T> __device__ int idx_from_t_xyz(int t, int xyz, T X[4])
```
We should put this into `index_helper.cuh`

maddyscientist replied:
3767371511982bbe8a6ec4cb23a09dda501b6272

weinbe2 commented on lib/communicator_qmp.cpp:
```diff
@@ -166,6 +166,9 @@ void Communicator::comm_init(int ndim, const int *dims, QudaCommsMap rank_from_c
               grid_size, QMP_comm_get_number_of_nodes(QMP_COMM_HANDLE));
   }
 
+  // defer handling MPI errors to QMP
```
I'm a bit confused, in `lib/communicator_mpi.cpp` an identical function call defers handling MPI errors to QUDA, why does this one defer it to QMP? Is this just a statement of "who called MPI_Init"?

maddyscientist replied:
I don't believe we have the capability to defer the QMP error handling to QUDA, though that would be desirable.  It's something that I will look into adding to QMP and then propagate enabling it into QUDA.

The motivation in the first place for this is to allow us to do post-error analysis, e.g., print a stack trace if `QUDA_BACKWARDS` is enabled, or anything else we might want to do if an MPI error is thrown.  I found this very helpful when encountering fatal MPI errors.

weinbe2 commented on tests/host_reference/staggered_dslash_reference.cpp:
```diff
@@ -29,12 +29,15 @@
  * @param[in] daggerBit Perform the ordinary dslash (0) or Hermitian conjugate (1)
  * @param[in] dslash_type The type of Dslash operation
+ * @param[in] laplace3D Whether we applying the 3-d laplace operator
+ * (in the case of dslash_type being QUDA_LAPLACE_DSLASH
```
missing a closing `)`

maddyscientist replied:
196cef438

weinbe2 commented on tests/host_reference/staggered_dslash_reference.cpp:
```diff
@@ -169,24 +171,21 @@ void stag_mat(ColorSpinorField &out, const GaugeField &fat_link, const GaugeFiel
 
   if (dslash_type == QUDA_LAPLACE_DSLASH) {
-    double kappa = 1.0 / (8 + mass);
+    double kappa = 1.0 / ((laplace3D == 3 ? 6 : 8) + mass);
```
iirc `laplace3D` is an integer equal to the dimension being partitioned, so in the off-chance someone is partitioning the `z` direction instead of `t` direction, `laplace3D` would be `2`, and this would be wrong.

maddyscientist replied:
196cef438

weinbe2 commented on tests/host_reference/staggered_dslash_reference.h:
```diff
@@ -23,7 +23,7 @@ void setDims(int *);
  * @param[in] dslash_type Dslash type
  */
 void stag_dslash(ColorSpinorField &out, const GaugeField &fat_link, const GaugeField &long_link,
```
Needs updated doxygen

weinbe2 replied:
This applies to multiple functions in this file

maddyscientist replied:
196cef438

weinbe2 commented on tests/staggered_eigensolve_test.cpp:
```diff
@@ -192,24 +192,75 @@ std::vector<double> eigensolve(test_t test_param)
+    quda::host_timer_t host_timer;
+    host_timer.start(); // start the timer
+
+    QudaGaugeObservableParam *obs_param = new QudaGaugeObservableParam[gauge_smear_steps / measurement_interval + 1];
```
This needs a corresponding `delete[]`

weinbe2 replied:
Or we could use a `std::vector<QudaGaugeObservableParam>` instead, I believe?

maddyscientist replied:
196cef438

weinbe2 commented on tests/staggered_eigensolve_test.cpp:
```diff
@@ -192,24 +192,75 @@ std::vector<double> eigensolve(test_t test_param)
+    // We here set all the problem parameters for all possible smearing types.
+    QudaGaugeSmearParam smear_param = newQudaGaugeSmearParam();
+    setGaugeSmearParam(smear_param);
+    smear_param.smear_type = gauge_smear_type;
```
Can (some of) these parameters be set in `setGaugeSmearParam`?

maddyscientist replied:
196cef438

weinbe2 commented:
I know that historically `laplace3D` has been an integer that in principle can be set to `0`, `1`, `2`, or `3` to specify the orthogonal direction, but iirc only the `t == 3` direction is only ever used (and implemented, in most cases). I also can't think of a real need for it to be set to anything but `3`. Should we just turn `laplace3D` into a `bool` and be done with it?

If someone has a need for any of the spatial dimensions to be the orthogonal direction, we can revisit the decision then. Getting the spatial dimensions working will be far more complicated of an enterprise than just flipping a switch anyway.

weinbe2 commented:
Can confirm the MILC `develop` safely compiles with the updated `quda.h`. I verified visually that it includes `-std=c99` in its default flags for compiling `ks_imp_rhmd` and `ks_spectrum`.

weinbe2 commented on include/kernels/blas_3d.cuh:
```diff
@@ -0,0 +1,341 @@
+#include <quda_matrix.h>
+#include <matrix_field.h>
+#include <kernel.h>
+//#include <kernels/contraction_helper.cuh>
```
This can be deleted

maddyscientist replied:
3767371511982bbe8a6ec4cb23a09dda501b6272

weinbe2 commented on include/kernels/blas_3d.cuh:
```diff
@@ -0,0 +1,341 @@
+    }
+  };
+
+  template <typename Arg> struct copyTo3d {
```
Should we be explicitly enforcing `x.X(3) == 1` within `copyTo3d` and `copyFrom3d`? I know there's a check in [lib/blas_3d.cu](https://github.com/lattice/quda/pull/1506/files#diff-1c272c738b087a6e48b33efe392437636acdfe9fabad2e4a1be03198a16ab4f4) , but maybe there's a straightforward way to do it here as well.

maddyscientist replied:
I don't think a check here is merited.  As you note there's already a check in the code that calls this in blas_3d.cu: one can get stuck ingoing deeper into wonderland if you don't trust anything at a given level.....

weinbe2 commented on include/blas_3d.h:
```diff
@@ -0,0 +1,35 @@
+     */
+    void swap(int slice, ColorSpinorField &x, ColorSpinorField &y);
+
+    // Reductions
```
doxygen

maddyscientist replied:
1d66a556b

weinbe2 commented on lib/gauge_laplace.cpp:
```diff
@@ -79,7 +79,10 @@ namespace quda {
-  GaugeLaplacePC::GaugeLaplacePC(const DiracParam &param) : GaugeLaplace(param) { }
+  GaugeLaplacePC::GaugeLaplacePC(const DiracParam &param) : GaugeLaplace(param)
+  {
+    for (auto i = 0; i < 4; i++) if (laplace3D == i) errorQuda("Cannot use 3-d operator with with e/o preconditioning");
```
Supurfluous `with`

maddyscientist replied:
c5b568171

weinbe2 commented on lib/solve.cpp:
```diff
@@ -394,18 +395,6 @@ namespace quda
       x[0] = ColorSpinorField(cudaParam);
     }
 
-    if (param.use_init_guess == QUDA_USE_INIT_GUESS_YES && !param.chrono_use_resident) { // download initial guess
```
Why are we deleting this chunk of code?

maddyscientist replied:
Accident as part of the merge of conflicted branches.  Thanks for catching this.

maddyscientist replied:
cbe9845685fc4fe72b878d68591377d2bc660f7a

weinbe2 commented on include/eigensolve_quda.h:
```diff
@@ -273,7 +276,7 @@ namespace quda
        @param[in] mat Matrix operator
        @param[in] evecs The eigenvectors
-       @param[in] evals The eigenvalues
+       @param[in,out] evals The eigenvalues
```
Now that we have a 3-d eigensolver, we should be make the `computeEvals` routine virtual and in the case of the 3-d eigensolvers make a custom override that `errorQuda`s out. This applies to many other routines as well---essentially anything that has a `3D` version in `TRLM3D`. We may also want to create a virtual function `virtual bool is_3d_eigensolver()` that by default returns `false` and for the 3-d eigensolvers returns `true`.

maddyscientist replied:
[1e2c0c7](https://github.com/lattice/quda/pull/1506/commits/1e2c0c7cf2da19395b92e18dcf4996a38087afd0) does something related: remove code replication and make `computeEvals3D` a specialization of the regular `computeEvals` which is what I was planning on doing anyway.

Marking as resolved, but we can discuss this more if needed.

weinbe2 commented on lib/eig_trlm_3d.cpp:
```diff
@@ -0,0 +1,701 @@
+        for (int i = 0; i < dim; i++) blas3d::copy(t, blas3d::COPY_TO_3D, vecs_t[i], vecs_locked[i]);
+
+        // Compute the axpy
+        blas::block::axpy(ritz_mat_keep[t], {vecs_t.begin(), vecs_t.begin() + dim},
```
Just confirming this is correctly `blas::` and shouldn't be `blas3d::`?

maddyscientist replied:
Yes, confirming this is correct.  We are using multi-blas routines on a pure 3-d grid here, which just works.  It's suboptimal, in that it would be better to have true batched 3-d multi-blas without the copy back and forward, but that will come in a subsequent PR.

weinbe2 commented on lib/eig_trlm_3d.cpp:
```diff
@@ -0,0 +1,701 @@
+  }
+
+  template <extremumType min_max, typename T>
+  T TRLM3D::getArrayMinMax3D(const std::vector<T> &array)
```
I think it could be worth splitting up this routine and making it a bona-fide member of `comm_quda.h`, have something like `comm_allreduce_array_min` and `comm_allreduce_array_max` or something of the sort. This could have far more value than just this context here (though I can't think of exactly what for right now).

maddyscientist replied:
We already have array min / max comm reduction functions, but that's not what's being used here.  What's here is simpler, and is a scalar reduction, not array reduction.  However, `getArrayMinMax3D` is completely unnecessary, and it's better to just delete it.  1ddda7c54 does this.

weinbe2 commented on lib/tune.cpp:
```diff
@@ -320,30 +347,26 @@ namespace quda
-   * @param[in] root_rank From which global rank to do the broadcast
-   */
-  static void broadcastTuneCache(int32_t root_rank = 0)
+  static void broadcastTuneCache(int32_t root_rank, map &tc_recv)
```
re-add doxygen?

weinbe2 replied:
nvm

weinbe2 commented on lib/tune.cpp:
```diff
@@ -152,9 +152,36 @@ namespace quda
+    }
+
+    // now merge the maps
+    for (auto i = 0u; i < global_tune_rank.size(); i++) { tunecache.merge(split_tc[i]); }
```
This needs to be modified to make the merge deterministic across ranks

weinbe2 replied:
After digging into the documentaiton on `std::map::merge`, it looks like this may not do what we want. Based on https://en.cppreference.com/w/cpp/container/map/merge , the `merge` routine doesn't do explicit copies: "No elements are copied or moved, only the internal pointers of the container nodes are repointed." This means that if/when `split_tc[i]` falls out of scope, its contents are invalidated, and as such the pointers to objects from `split_tc[i]` now in `tunecache` are also invalid.

A more concrete code example of my thought:

```
void add_things(std::map<...> &base) {
  std::map<...> local_thing;
  // populate local_thing
  base.merge(local_thing);
  // local_thing falls out of scope
}

...

std::map<...> blah = { ... };
// I want to add things to blah
add_things(blah);

// "blah" now, in theory, contains elements that point to elements of "local_thing", but since "local_thing" fell out of scope those pointers are now invalid
```

maddyscientist replied:
I've confirmed that `std::map::merge` does do what we want.  When `split_tc[I]` goes out scope there is no issue.  The text regarding `..the internal pointers of the container nodes are repointed...` concern the underlying container of the respective maps.  While a `std::map` doesn't use a (doubly) linked list (will use a red-black tree instead), it's instructive to think in these terms: we splice in various entries by updating the appropriate next and prev pointers of adjacent entries in the list we're inserting an entry into.  No data is moved, only the pointers updated.

weinbe2 commented:
Looking at the host verify... at this point I think we should be splitting the Laplace verification routines out from the staggered ones at this point, if anything for code cleanliness/maintainability reasons. Alternatively, there needs to be many more checks for invalid combinations: `matpc` and `matdagmat` don't make sense for the 3-d Laplace operator, `laplace3D` doesn't make sense for the staggered operators... I'm sure I could find more cases of inconsistencies if I wanted to.

maddyscientist committed:
Fix bad merge

weinbe2 commented on lib/blas_3d.cu:
```diff
@@ -0,0 +1,251 @@
+        if (x.Ncolor() != y.Ncolor()) errorQuda("Unexpected number of colors x=%d y=%d", x.Ncolor(), y.Ncolor());
+
+        // Check slice value
+        if (t_slice >= y.X()[3]) errorQuda("Unexpected slice %d", t_slice);
```
Extend to `< 0` as well

maddyscientist replied:
c5b568171

weinbe2 commented on lib/blas_3d.cu:
```diff
@@ -0,0 +1,251 @@
+      checkPrecision(x, y);
+
+      // Check spins
+      if (x.Nspin() != y.Nspin()) errorQuda("Unexpected number of spins x=%d y=%d", x.Nspin(), y.Nspin());
```
Maybe out of the scope of this PR, but either here or in the future (I can create an issue, let me know) we should create a set of routines `checkSpin`, `checkColor`, `checkSpinColor` that acts similarly to `checkPrecision`, etc.

maddyscientist replied:
Yeah, I would say that's out of scope, but certainly good to have.  Please create an issue 😉

maddyscientist replied:
I just went and did it dfd36e754

weinbe2 commented on lib/eig_trlm_3d.cpp:
```diff
@@ -0,0 +1,701 @@
+#include <vector>
+#include <algorithm>
+
+#include <quda_internal.h>
```
Should we start implementing using `#include "..."` for headers inside QUDA, `#include <...>` for external headers with this PR? Or should we just do a mega-change at some point in the future?

maddyscientist replied:
Let's defer that joy for another day....

weinbe2 commented on lib/eig_trlm_3d.cpp:
```diff
@@ -0,0 +1,701 @@
+      }
+
+      if (getVerbosity() >= QUDA_VERBOSE && comm_coord(0) == 0 && comm_coord(1) == 0 && comm_coord(2) == 0) {
+        printf("iter = %d rank = %d converged = ", restart_iter + 1, comm_rank_global());
```
I see what's going on here, do we want to make a custom `printfQuda` (`printfQuda3d`?) that handles this type of conditional on the print instead of using `printf`? Or should we avoid over-specializing?

In any case---it might be good to modify this to build the entire string before printing so the outputs from different ranks don't get interleaved... or at least minimize the risk.

maddyscientist replied:
I'd steer away from `printfQuda3d` but your point about building the entire string first is a good one.

maddyscientist replied:
030775807

weinbe2 commented on lib/eig_trlm_3d.cpp:
```diff
@@ -0,0 +1,701 @@
+    getProfile().TPSTART(QUDA_PROFILE_EIGEN);
+
+    // Loop over the 3D problems
+#ifdef _OPENMP
```
Is there a particular need for this `#ifdef` or can we just get rid of it?

weinbe2 replied:
To be clear I understand the purpose of OpenMP parallelism here, my only question is about the guard around it

maddyscientist replied:
Indeed the guard isn't needed, will be nuked.

maddyscientist replied:
c5b568171

weinbe2 commented on lib/eig_trlm_3d.cpp:
```diff
@@ -0,0 +1,701 @@
+
+  void TRLM3D::prepareInitialGuess3D(std::vector<ColorSpinorField> &kSpace, int ortho_dim_size)
+  {
+    if (sqrt(blas::norm2(kSpace[0])) == 0.0) {
```
I learned the hard way through interfacing with MILC that we should also check for +/- inf and nan in the context of completely uninitialized data, see here: https://github.com/lattice/quda/blob/develop/lib/eigensolve_quda.cpp#L145-L150

maddyscientist replied:
Good call.  Will use `std::isfinite` here.

maddyscientist replied:
c5b568171

cjmorningstar10 commented:
In  lib/interface_quda.cpp   on line 2577
         GaugeField *cudaGauge = checkGauge(inv_param);
For the 3d laph ev calculation, if use_smeared_gauge is TRUE,  the original gauge configuration is not needed, only the smeared gauge is needed, so the above check can fail if prior tasks do not load the original config.   In this case, only a check of the smeared gauge field is needed.

maddyscientist committed:
Add doxygen to blas-3d functions

maddyscientist committed:
Move idx_from_t_xyz to index_helper.cuh

maddyscientist committed:
Apply suggestions from PR review

maddyscientist committed:
Apply PR review comments relevant to tests dir

maddyscientist committed:
Remove redundant wilson distance preconditioned split grid test

maddyscientist committed:
Fix split grid ctest command

maddyscientist committed:
Remove unnecessary TRLM specializations, and make computeEvals3D a specialization of computeEvals

maddyscientist committed:
Fix CI

maddyscientist committed:
Remove unnecessary getArrayMinMax3D function

maddyscientist committed:
Some verbosity cleanup for TRLM-3d

maddyscientist committed:
Add checkColor and checkSpin helpers

maddyscientist committed:
Split grid tunecache merge is now deterministic across sub grids

maddyscientist committed:
In host reference code check we don't request e/o preconditioning with 3-d Laplace operator, and check we don't use 3-d operator when using staggered

maddyscientist committed:
checkGauge will only check smeared guage field if use_smeared_gauge = true

maddyscientist commented:
> In lib/interface_quda.cpp on line 2577 GaugeField *cudaGauge = checkGauge(inv_param); For the 3d laph ev calculation, if use_smeared_gauge is TRUE, the original gauge configuration is not needed, only the smeared gauge is needed, so the above check can fail if prior tasks do not load the original config. In this case, only a check of the smeared gauge field is needed.

Thanks @cjmorningstar10, 3ae79cc61 takes care of this.

maddyscientist commented:
> Looking at the host verify... at this point I think we should be splitting the Laplace verification routines out from the staggered ones at this point, if anything for code cleanliness/maintainability reasons. Alternatively, there needs to be many more checks for invalid combinations: `matpc` and `matdagmat` don't make sense for the 3-d Laplace operator, `laplace3D` doesn't make sense for the staggered operators... I'm sure I could find more cases of inconsistencies if I wanted to.

Given the similarity between the staggered and Laplace kernels (they are essentially identical), I would push back on this, at least in the scope of this PR.  I have, however, added the consistency checks you suggest (0e906a6d89c4bb73659ac99c9e9cd002f93dc1df).

maddyscientist committed:
Apply clang-format

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into feature/trlm_3d

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into feature/trlm_3d

maddyscientist committed:
Some cleanup of Lapalce operator: remove redundant dagger

maddyscientist committed:
Fix some typos

maddyscientist committed:
Fix Laplace solver tests (reference Laplace operator uses kappa normalization, but staggered invert test was hard coded to use mass normalization).  Fix split grid ctest when using single GPU

maddyscientist committed:
Add 3-d Laplace TRLM-3d to ctest

maddyscientist commented:
@weinbe2 I believe your points are addressed, and I have added the final few things I wanted to get done: mainly adding Laplace TRLM-3d to the ctest list.

weinbe2 commented on include/blas_3d.h:
```diff
@@ -0,0 +1,80 @@
+  {
+
+    // Local enum for the 3D copy type
+    enum copy3dType { COPY_TO_3D, COPY_FROM_3D, SWAP_3D };
```
I've been trying to use `enum class` for internal (i.e., non-`quda_enum.h`) enums, I think we should apply that here.

maddyscientist replied:
[6383489](https://github.com/lattice/quda/pull/1506/commits/63834893048dbee7acee91f788bbc14003d933cc)

maddyscientist committed:
Switch to enum class for blas3d::copyType

weinbe2 commented on include/color_spinor_field.h:
```diff
@@ -1054,6 +1056,38 @@ namespace quda
+     @param[in] b Input field
+     @return If spin is unique return the number of spins
+   */
+  inline int Spin_(const char *func, const char *file, int line, const ColorSpinorField &a, const ColorSpinorField &b)
```
Should we add a version of this for vector_ref/cvector_ref/etc of this as well?

maddyscientist replied:
I would say no.  The `Nspin()` methods of `cvector_ref<ColorSpinorField>` themselves already test internal for consistency.  I don't think we need to go the length of comparisons between arbitrary combinations of fields and cvector_refs of fields.

weinbe2 commented on lib/communicator_stack.cpp:
```diff
@@ -49,8 +50,16 @@ namespace quda
+      errorQuda("Split-grid is currently not supported with NVSHMEM. Set QUDA_ENABLE_NVSHMEM=0 to disable NVSHMEM.");
+
+    // if reverting to global, we will need to join the tunecaches
+    bool join_tune_cache = split_key == default_comm_key;
```
Shouldn't we theoretically need to do this each time we go to a `split_key` with a smaller `[key].product()`?

maddyscientist replied:
Sorry, not following.  What do you mean?

weinbe2 replied:
If there are recursive split-grid calls, i.e., further splitting an already split grid.

maddyscientist replied:
Ah, gotcha now.  That is precluded and recursive split grids are not supported.

maddyscientist replied:
The way to do it would be with a `std::stack` I guess.

maddyscientist committed:
Increase max local size of blas-3d batched kernels to 256

weinbe2 commented:
I think we should add a test of the 3-d reduction routines to `blas_test`, but in the interest of getting this PR merged I'm okay with that being added to a later PR, just file an issue about it :)

maddyscientist commented:
> I think we should add a test of the 3-d reduction routines to `blas_test`, but in the interest of getting this PR merged I'm okay with that being added to a later PR, just file an issue about it :)

That sounds good.  Eventually I'm going to nuke these routines entirely with them to be replaced by the domain-decomposed bias that I've threatened to do on the Domain Decomposition calls.

weinbe2 commented on tests/utils/host_blas.cpp:
```diff
@@ -63,20 +63,20 @@ void mxpy(void *x, void *y, int len, QudaPrecision precision)
   double sum = 0.0;
   for (int i = 0; i < len; i++) sum += v[i] * v[i];
-  quda::comm_allreduce_sum(sum);
+  if (global) quda::comm_allreduce_sum(sum);
```
I need to think about it a bit, but I think that for the context this is used in (3-d laplace eigensolvers) you don't want to outright disable the full reduction, you only want to reduce in the `x/y/z` plane?

maddyscientist replied:
The disablement of the global reduction was intentional here, as the point was to defer the global reduction until later and do it once in a padded structure to allow for a single 4-d all reduce to get the correct answer.  Looks like I didn't actually do this properly so I'll fix this.

maddyscientist replied:
38b1610d52b2503ce467ed1a706e394eda8d8fa1

weinbe2 commented on tests/utils/set_params.cpp:
```diff
@@ -130,11 +134,11 @@ void setInvertParam(QudaInvertParam &inv_param)
     inv_param.mass = mass;
     inv_param.kappa = 1.0 / (2.0 * (1 + 3 / anisotropy + mass));
-    if (dslash_type == QUDA_LAPLACE_DSLASH) inv_param.kappa = 1.0 / (8 + mass);
+    if (dslash_type == QUDA_LAPLACE_DSLASH) inv_param.kappa = 1.0 / ((laplace3D == 3 ? 6 : 8) + mass);
```
I saw elsewhere in the code that `(laplace3D == 3)` was replaced with something more like `(laplace3D < 4)` or `(laplace3D != -1)` or something like that, I forget the convention; I preferred that to the hard compare against `3`, and it seems safer.

maddyscientist replied:
The less than convention was there to give the option for future X/Y/Z orthogonal dimension.  Good to make it consistent though.

maddyscientist replied:
8fb71814ce06ce0a626b6a7c3ec7deae024e054a

maddyscientist committed:
Fix compilation when large arg is disabled

weinbe2 reviewed (approved):
Compiling against MILC looks good, as does testing MG. This is good to go (pending reconciling the testing failures with the latest changes). Awesome work @maddyscientist and @cpviolator

maddyscientist committed:
Fix some issues with reference Laplace3D eigenvector residual computation on multiple GPUs

maddyscientist committed:
Cleanup verbosity handling in TRLM-3D

maddyscientist committed:
Cleanup handling of laplace3D when computing kappa in reference code

maddyscientist committed:
clang-format

maddyscientist commented:
We should be good now for this PR.  We can merge this once CI completes 🥳

maddyscientist commented:
cscs-ci run

maddyscientist committed:
Revert changes to checkGauge
